DOCSP-45153 Add MONGODB-OIDC (#152)
mongodb/docs-rust: source/fundamentals/enterprise-auth.txt

   
   let uri = "mongodb://<username>:<password>@<hostname>/?authSource=$external&authMechanism=PLAIN";

.. _rust-mongodb-oidc:

MONGODB-OIDC
------------

.. important::

   The MONGODB-OIDC authentication mechanism requires MongoDB Server
   v7.0 or later running on a Linux platform.

The {+driver-short+} supports OpenID Connect (**OIDC**) authentication for **workload
identities**. A workload identity is an identity you assign to a
software workload, such as an application, service, script, or
container, to authenticate and access other services and resources.

The following sections describe how to use the MONGODB-OIDC
authentication mechanism to authenticate to various platforms.

To learn more about the MONGODB-OIDC authentication mechanism, see
:manual:`OpenID Connect Authentication </core/security-oidc/>` and
:manual:`MongoDB Server Parameters </reference/parameters/#mongodb-parameter-param.oidcIdentityProviders>`

> in the server manual.

Author: I noticed in the Rust docs, other pages just say "Server manual" so I opted to just capitalize server here just to stay consistent and avoid touching a lot of other lines with this PR, but will check with the main repo steward to see how we want to use this going forward.

.. _rust-mongodb-oidc-azure-imds:

Azure IMDS
~~~~~~~~~~

If your application runs on an Azure VM, or otherwise uses the
`Azure Instance Metadata Service <https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service>`__
(IMDS), you can authenticate to MongoDB by using the {+driver-short+}'s
built-in Azure support.

You can configure OIDC for Azure IMDS by setting the ``mechanism`` field of your
``Credential`` struct to ``AuthMechanism::MongoDBOidc``. The following example

> specifies the authentication mechanism by using the following placeholders:

Reviewer: S: feels cleaner to avoid using "following" twice (i know this is tough to avoid). could you do something similar to what you did in the next example (lines 173-183)?

- ``username``:  If you're using an Azure managed identity, set this to the
  client ID of the managed identity. If you're using a service principal to
  represent an enterprise application, set this to the application ID of the
  service principal.
- ``mechanism_properties``: Set the ``ENVIRONMENT`` property to ``azure`` and

>   the ``TOKEN_RESOURCE`` to with the value of the audience parameter configured

Reviewer: ```suggestion
  the ``TOKEN_RESOURCE`` to the value of the audience parameter configured
```

  on your MongoDB deployment.

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-azure-imds
   :end-before: end-azure-imds
   :emphasize-lines: 2-5

.. _rust-mongodb-oidc-gcp-imds:

> GCP IMDS

Reviewer: Q: does the rust driver natively support gcp?

Author: Yep found out that it does, I got an example from Kevin via Slack 👍

~~~~~~~~

If your application runs on a Google Compute Engine VM, or otherwise uses the
`GCP Instance Metadata Service <https://cloud.google.com/compute/docs/metadata/querying-metadata>`__,
you can authenticate to MongoDB by using the {+driver-short+}'s built-in GCP
support.

You can configure OIDC for GCP IMDS by setting the ``mechanism`` field of your
``Credential`` struct to ``AuthMechanism::MongoDBOidc``. Then, specify the
authentication mechanism by setting the following values in the
``mechanism_properties`` field:

- ``ENVIRONMENT``: Set this property to ``gcp``.
- ``TOKEN_RESOURCE``: Set this property to the value of the audience parameter configured
  on your MongoDB deployment.

The following code example shows how to set these options when creating a
``Client``:

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-gcp-imds
   :end-before: end-gcp-imds
   :emphasize-lines: 2-4

.. _rust-mongodb-oidc-custom-callback:

Custom Callback
~~~~~~~~~~~~~~~

The {+driver-short+} doesn't offer built-in support for all platforms,

> including the AWS Elastic Kubernetes Service (EKS). To authenticate
> against unsupported platforms, you must define a custom callback
> function to use OIDC to authenticate.

Reviewer: S:

```suggestion
including the AWS Elastic Kubernetes Service (EKS). To use OIDC to authenticate
against unsupported platforms, you must define a custom callback
function.
```

The following example defines a custom callback for an EKS cluster with a
configured Identity and Access Management (IAM) OIDC provider. The access token
is read from a path set in the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment
variable. Then, you can set the ``oidc_callback`` field of your ``Credential``

> struct to ``oidc::Callback::machine``:

Reviewer: S: present this as a series of steps:

First, set the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable...then, set this field, as shown in the following example:

Author: Rephrased this a bit and added some additional relevant resources.

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-custom-callback-machine
   :end-before: end-custom-callback-machine
   :emphasize-lines: 3, 5-10

When the workforce identity authentication process involves human interaction,
you must configure the client by setting the ``oidc_callback`` field of your
``Credential`` struct to ``oidc::Callback::human`` instead of
``oidc::Callback::machine``. The {+driver-short+} uses the callback in the
following process:

1. The driver retrieves the Identity Provider Information (IDPInfo) for the
   provided username.

> 2. The callback negotiates with the IDP to obtain an ``AccessToken``, possible
>    ``RefreshToken``, and any timeouts, then returns them.

Reviewer: ```suggestion
2. The callback negotiates with the IDP to obtain an ``AccessToken`` and
     any timeout values, then returns them. It might also return a 
    ``RefreshToken`` (I assume this depends on server config? If so, I'd mention this)
```

``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable. Finally, set the value of
the ``access_token`` field of the ``IdpServerResponse`` struct. Optionally, set
the values of the ``expires`` and ``refresh_token`` fields.

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-custom-callback-machine
   :end-before: end-custom-callback-machine
   :emphasize-lines: 3, 5-7

When the workforce identity authentication process involves human interaction,
you must configure the client by setting the ``oidc_callback`` field of your
``Credential`` struct to ``oidc::Callback::human`` instead of
``oidc::Callback::machine``. The {+driver-short+} uses the callback in the
following process:

1. The driver retrieves the Identity Provider Information (IDPInfo) for the
   provided username.
#. The callback negotiates with the IDP to obtain an ``access_token``, and any
   potential ``refresh_token`` and timeout values, if configured, then returns
   them.

The following example defines a custom callback to handle workforce identity. To
customize this example for your use case, replace ``<human flow>`` with your own

> custom flow. Refer to `Authorization Code Flow with OIDC

Reviewer 2: it might be worth mentioning here that the callback should return a `Result<IdpServerInfo>` since that's not shown explicitly in the code example

Author: Updated the explainer paragraph before the code snippet and added an API docs link to the resources at the bottom of the guide.
